### ISSUE TITLE
level.dat Biome Rules Recurse Infinitely [Problem Area Identified, just needs fix]

### ISSUE THREAD

**Issue by ByThePowerOfScience** [AlexModGuy/Citadel#192]
Seems to be the exact same issue as #101, but on the latest version of Citadel.  Happening consistently with this pack on all platforms, and both client and server-side.

I spent like twelve hours straight going through it with a debugger trying to find why it was doing that, so if you need any more information, please let me know.

Citadel: citadel-2.6.0-1.20.1
TerraBlender: Disabled

[level.dat.txt (an extension that GitHub will accept)](https://github.com/user-attachments/files/17685999/level.dat.txt)
![NBTExplorer_2024-11-09_00-00-37](https://github.com/user-attachments/assets/ab38f8bb-0ffd-4947-9c8e-5691140b704f)

**Comment by ByThePowerOfScience** [comment]
It looks like it might be an incompatibility with [Luminous: Nether](https://www.curseforge.com/minecraft/mc-mods/luminous-nether/), though how an MCreator mod could do that is beyond me.

**Comment by ByThePowerOfScience** [comment]
Removing Luminous: Nether did not, in fact, fix it.  However, I think I found the cause.

In `mixin.NoiseBasedChunkGeneratorMixin`, the biome rules are appended to the existing biome rules.  However, this does not take into account the fact that the previously injected biome rules were actually already deserialized and exist there already.  So it just keeps adding them onto the end of the list, forever.

No mod incompatibility, just a Citadel bug because the `mergedSurfaceRules` flag doesn't persist between restarts.

**Comment by kinnikinnick** [comment]
So is this what's causing the worldgen to be so incredibly slow after playing on the server awhile?  Is there a fix for this aside from removing Citadel and all mods that require it?

**Comment by Veynam** [comment]
Id like to know as well. Ive been searching for a solution for the past 2 months. Frustrating

**Comment by ByThePowerOfScience** [comment]
> Id like to know as well. Ive been searching for a solution for the past 2 months. Frustrating

> So is this what's causing the worldgen to be so incredibly slow after playing on the server awhile? Is there a fix for this aside from removing Citadel and all mods that require it?

Until the issue is resolved, the only fix is to open up `level.dat` in NBTExplorer and remove the infinitely recursing rule.

**Comment by soup587** [comment]
what exactly would the fix be here? storing mergedSurfaceRules somewhere that gets saved across restarts?

**Comment by ByThePowerOfScience** [comment]
> what exactly would the fix be here? storing mergedSurfaceRules somewhere that gets saved across restarts?

Yeah, pretty much.

Alternately, use MixinExtras' `@ModifyReturnValue` to allow compat with Terrablender without saving state, instead of modifying `this.surfaceRule`.

**Comment by ByThePowerOfScience** [comment]
@AlexModGuy I figured it out!!! I've been having this issue for a solid year now and I keep trying to fix it with Mixins, but I finally figured it out!!!!!

# Problem

The problem is this: every time the `level.dat` is saved, the new biome rules rewrap the old ones that already include the biome rules wrapping the old ones, causing infinite nesting resulting in an eventual crash.  

The _reason_ this happens is that we're modifying the _getter_, **which is the same thing used by the _codec_ when serializing it**!  Even _my_ change that uses `@ModifyReturnValue` _doesn't fix anything_, because we're still _re-adding the biome rules each and every time the game is run_!

# Solution

There are two ways to fix this: 

## Possibility 1

Add persistent data to the Level (aka just use WorldData) to tell us that we've already added our biome rules.  

The pro of this is that it's much faster, as it's all cached.  We can still freely modify the `surfaceRule` field to cache the calculation when we add them, but we have to do it _once and only once ever_.

However, the issue _now_ becomes "what if we need to add new biome rules later? How do we tell what we've added when we're wrapping it again?"  This will require a datafixer to visit and patch the level.dat, meaning we'll have to figure out a way to identify our injected biome rules in the midst of the cluster of random biome rules.

## Possibility 2

Inject into `LevelStorageSource` and set a global flag so we know when the level.dat is being saved.  If it's being saved, don't add our biome rules.  This keeps the biome rule addition solely locked to runtime, removing the possibility of nesting.

The problem now is that we by necessity can't cache the additional biome rules, instead adding them each and every time the method is called.  This slows down chunk generation a not-inconsiderable amount.

...

I'm not sure which is the better option, but the current situation is untenable.  I'm currently having to manually un-brick my world every few days or so, so we've got to do something.

**Comment by AlexModGuy** [comment]
@ByThePowerOfScience Thank you for helping figure this one out!
I may have a tentative solution, but it'll require some testing to figure out.
https://github.com/AlexModGuy/Citadel/blob/dd42eb2cfd1d0dc1c813d8f642c3e20063322d95/src/main/java/com/github/alexthe666/citadel/mixin/NoiseGeneratorSettingsMixin.java#L31
I've reverted back to 2.5.x's solution of modifying the surfaceRule in NoiseGeneratorSettings rather than NoiseBasedChunkGenerator for simplicity. I've added a check that compares a seed value generated by a hashed string of all of the new sequenced rules to be added. Hopefully, this code will only run on the occasions:
- level creation
- surface rule seed checksum changed (add/removed mods etc).

As I've said, still requires some testing. I should be able to have a Jar soonish, so if that's something you would like to test yourself hands-on, you can contact me on discord for that.

**Comment by AlexModGuy** [comment]
> [@ByThePowerOfScience](https://github.com/ByThePowerOfScience) Thank you for helping figure this one out! I may have a tentative solution, but it'll require some testing to figure out.
> 
> [Citadel/src/main/java/com/github/alexthe666/citadel/mixin/NoiseGeneratorSettingsMixin.java](https://github.com/AlexModGuy/Citadel/blob/dd42eb2cfd1d0dc1c813d8f642c3e20063322d95/src/main/java/com/github/alexthe666/citadel/mixin/NoiseGeneratorSettingsMixin.java#L31)
> 
> Line 31 in [dd42eb2](/AlexModGuy/Citadel/commit/dd42eb2cfd1d0dc1c813d8f642c3e20063322d95)
>  if (citadelServerData != null && !citadelServerData.isUsingLatestSurfaceRules()) { 
> 
> I've reverted back to 2.5.x's solution of modifying the surfaceRule in NoiseGeneratorSettings rather than NoiseBasedChunkGenerator for simplicity. I've added a check that compares a seed value generated by a hashed string of all of the new sequenced rules to be added. Hopefully, this code will only run on the occasions:
> 
>     * level creation
> 
>     * surface rule seed checksum changed (add/removed mods etc).
> 
> 
> As I've said, still requires some testing. I should be able to have a Jar soonish, so if that's something you would like to test yourself hands-on, you can contact me on discord for that.

![Image](https://github.com/user-attachments/assets/52534f80-f154-4ea9-ab12-743f2059c289)
Appears to have the unwanted side effect of breaking surface rules on relog for client worlds. Shame, I'll have to try something else.

**Comment by AlexModGuy** [comment]
Possible solution is reverting to the old way of merged rules, with some needed changes:
- wrap all surfaces rules added by citadel with `CitadelSurfaceRuleWrapper`
- before merging surface rules, [strip all `CitadelSurfaceRuleWrapper` surface rules](
https://github.com/AlexModGuy/Citadel/blob/1.20/src/main/java/com/github/alexthe666/citadel/server/generation/SurfaceRulesManager.java#L76).
Will still need to investigate and see if this fixes it or not.

**Comment by ByThePowerOfScience** [comment]
I just thought of another alternative: you could wrap your injected surface rules in a custom tag type with its own identifier, then use a Mixin to make that custom class be recognized as vanilla surface rules.

**Comment by KlemenDEV** [comment]
@ByThePowerOfScience has reported https://github.com/MCreator/MCreator/issues/6088, where I did some research, and here are some findings.

----

If mods work correctly, serialization of surface rules should not happen. When surface rules are serialized for the overworld, this creates compatibility issues if the mod is removed or biomes change (see https://github.com/MCreator/MCreator/issues/5499). There are two ways to override surface rules:

1. Use AT to replace the NoiseGeneratorSettings with a custom implementation that has a custom surfaceRule
2. Use Mixin to override surfaceRule() getter method

If any mod replaces NoiseGeneratorSettings, they will be serialized from the ground up instead of using the vanilla reference.

If mods use mixin to override surfaceRule() return value, this is done "on the fly" each time, and no risk of incompatibilities arises.

### If any mod overrides NoiseGeneratorSettings for overworld

<img width="465" height="427" alt="Image" src="https://github.com/user-attachments/assets/ab1acfe9-ec91-456a-bc4a-0a2969cb3ab6" />

### If all mods use mixin to override surfaceRule() getter method

(we can see only a reference to vanilla overworld NoiseGeneratorSettings is made)

<img width="1021" height="827" alt="Image" src="https://github.com/user-attachments/assets/6b10e325-87bd-47e1-b9eb-fd165007184a" />

----

This means the actual issue reported here was caused by neither Citadel nor [Herios' Floral Expansion](https://www.curseforge.com/minecraft/mc-mods/herios-floral-expansion).

The fix proposed at https://github.com/AlexModGuy/Citadel/issues/192#issuecomment-3013362545 is actually not needed, as this issue with cascading surface rules will always happen as long as at least one mod in the modpack replaces the actual NoiseGeneratorSettings, and in this case, the risk of cascading in level.dat exists.

----

For the case of @ByThePowerOfScience, the mod that caused this was shineals_prehistoric_expansion-1.4.3-neoforge-1.21.1.jar, which was made using an older version of MCreator, before this was fixed in MCreator. See https://github.com/MCreator/MCreator/issues/5499 and the fix at https://github.com/MCreator/MCreator/pull/5528

So the true solution when this happens is not making sure to return the unmodified surfaceRule() return value during world save serialization, but to not have mods override NoiseGeneratorSettings.

The only known source of mods that use(d) NoiseGeneratorSettings override was old MCreator mods. If such a mod is found, the author needs to be asked to update MCreator and re-export the mod.

----

Disclaimer: I am the maintainer of MCreator

We made a mistake of not using a mixin and instead replacing NoiseGeneratorSettings, and thus not being compatible with Terrablender and/or Citadel. We have fixed this issue as soon as it was reported at https://github.com/MCreator/MCreator/issues/5499, but some users have not updated MCreator yet or use outdated plugins, and such mods may appear from time to time. The only solution is to ask the mod author to update, because all mods using old MCreator versions will force serialization.

The issue was fixed almost a year ago in MCreator EAP 2025.2.24711 (https://mcreator.net/changelog), so this version or newer no longer directly overrides the field in NoiseGeneratorSettings:

> [Bugfix] MCreator mods in combination with Terrablender broke worlds after they were loaded 512 times

**Comment by ByThePowerOfScience** [comment]
@KlemenDEV What I reported for MCreator seems like a separate issue to the one here.  When I offered my fix, I did my due diligence and traced through the code, and confirmed that indeed the codec is used to serialize it to disk, at least in 1.20.1 where this issue was reported.

~~I think the reason your version doesn't cause issues is because you only modify the return value of the getter if the getter hasn't already been invoked, meaning in certain cases your rules aren't even being used for biome generation, but they definitely aren't injected when serialized to disk.~~

EDIT: Nevermind, I misremembered the MCreator code.  Thing is, I distinctly remember the world loading process using the NoiseGeneratorSettings codec to save the data.  Like that's a thing I confirmed to happen.

**Comment by KlemenDEV** [comment]
I have tested with your modpack in 1.21.1 (the version you provided for testing), and removing the mod that replaced NoiseGeneratorSettings fixed the issue as NoiseGeneratorSettings were no longer serialized, as they shouldn't, as explained above.

> but they definitely aren't injected when serialized to disk.

In 1.21.1, serialization of the overworld NoiseGeneratorSettings is not supposed to happen. A different story for custom dimensions, for but vanilla overworld, only reference is stored.

----

Now in 1.20.1, we don't support this version anymore. The last supported version of 1.20.1 in the MCreator core (2 years ago) still did the NoiseGeneratorSettings replacement, causing serialization.

I am not sure how the 1.20.1 plugin handles this, but I will forward the issue to the plugin author for further investigation.
